architecture/2025-02-application-autoscaling.md

```diff
@@ -0,0 +1,239 @@
+
+#### Option 1: Adding autoscaling in the compute configuration
+
+This enables to configure compute specific autoscaling policies. For e.g. : If the platform is Kubernetes, the default HPA config can be added to the runtime configuration and for serverless platform, the default HTTP based autoscaling config can be added to the runtime configuration.
+
+Eg : Kubernetes HPA config
+
+```bicep
+resource container 'Applications.Containers@2023-10-01-preview' = {
+  name: 'myapp'
+  properties: {
+    image: 'myregistry.azurecr.io/myapp:latest'
+    ports: {
+    }
+    runtimes: {
+      kubernetes: {
+        autoscaling: {
+          minReplicas: 1
+          maxReplicas: 10
+          cpuUtilization: 50
+          memoryLimit: 50      
+        }
+      }
+    }
+  }
+}
+```
+
+Eg : ACA HTTP based autoscaling config
+
+```bicep
+resource container 'Applications.Containers@2023-10-01-preview' = {
+  name: 'myapp'
+  properties: {
+    image: 'myregistry.azurecr.io/myapp:latest'
+    ports: {
+    }
+    runtimes: {
+      serverless-platforms: {
+        autoscaling: {
+          minReplicas: 1
+          maxReplicas: 10
+          http-concurrency: 50  
+        }
+      }
+    }
+  }
+}
+```
+
+Pros: 
+1. Easy to understand and configure.
+1. Autoscale config is run-time specific and doesn't have to be standardized across runtimes as different platforms have different autoscaling mechanisms. 
+1. Enterprises pick platforms based on their application workloads and the underlying platform provides default autoscaling policies based on the workloads. Radius provides a clear abstraction to leverage default autoscaling mechanisms from the underlying platform.
+
+Cons:
+1. More configuration to manage across different runtimes.
+2. The developer needs to know the underlying platform and the autoscaling policies required for the platform.
+
+#### Option 2: Adding it as part extensions
+
+`extensions` have been the way to provide punch through capabilities which do not change the behaviors of the resource. The following is an example of adding autoscaling as part of the extensions. 
+
+```bicep
+resource container 'Applications.Containers@2023-10-01-preview' = {
+  name: 'myapp'
+  properties: {
+    image: 'myregistry.azurecr.io/myapp:latest'
+    extensions: [
+      {
+        kind: 'autoscaling'
+        minReplicas: 1
+        maxReplicas: 10
+        cpuUtilization: 50
+        memoryLimit: 50
+        httpConcurrency: 50
+        }
+      }
+    ]
+  }
+}
+```
+
+Pros: consistent with existing manual scaling config
+
+Cons: not straightforward or intuitive as user doesn't know which autoscale config is required for the platform ? needs a platform specific discriminator
+May be, this is not a developer problem and should be handled by the platform operator. 
+
+#### Option 3: Adding it as top level common property
```

Comment:
I think Option 3 makes the most sense IF hpa and/or http auto-scaling and their corresponding properties are common across container platforms.